pkgs/applications/networking/cluster/terraform-providers/default.nix

```diff
@@ -2,11 +2,16 @@
 , buildGoPackage
 , fetchFromGitHub
 , callPackage
+, runtimeShell
 }:
 let
-  list = import ./data.nix;
+  list = lib.importJSON ./providers.json;
 
-  toDrv = data:
+  toDrv = name: data:
+    let
+      fallbackProviderSourceAddress = "nixpkgs/${data.owner}/${name}";
```

Comment:
Provide a fallback for providers which haven't updated to use the Terraform registry. But I haven't able to make this work.
```tf
# main.tf
terraform {
  required_providers {
    acme = {
      source = "getstackhead/acme"
      version = "1.5.0-patched"
    }
    auth0 = {
      source = "nixpkgs/terraform-providers/auth0"
      version = "0.11.0"
    }
  }
}

provider "acme" {
}

provider "auth0" {
}
```

```bash
nix-shell -I nixpkgs=$PAHT_TO_NIXPKGS -p 'terraform_0_13.withPlugins (p : [p.acme p.auth0])' --command "terraform init"

Initializing the backend...

Initializing provider plugins...
- Using previously-installed getstackhead/acme v1.5.0-patched
- Finding nixpkgs/terraform-providers/auth0 versions matching "0.11.0"...

Error: Failed to query available provider packages

Could not retrieve the list of available versions for provider
nixpkgs/terraform-providers/auth0: provider nixpkgs/terraform-providers/auth0
was not found in any of the search locations

- /nix/store/04xgzlpfc1slds41idrh5j81l8h85mpg-tf-plugin-env/bin

```

And yet
```
tree -l /nix/store/04xgzlpfc1slds41idrh5j81l8h85mpg-tf-plugin-env/bin
/nix/store/04xgzlpfc1slds41idrh5j81l8h85mpg-tf-plugin-env/bin
├── nixpkgs -> /nix/store/y16ks7jgads2y8gjzv6y7dv5sd6ydnln-terraform-provider-auth0-0.11.0/bin/nixpkgs
│   └── terraform-providers
│       └── auth0
│           └── 0.11.0
│               └── linux_amd64
│                   └── terraform-provider-auth0_v0.11.0
├── registry.terraform.io -> /nix/store/fay4phh1p79wbl4ywzf9zcvs5hcs4sf4-terraform-provider-acme-1.5.0-patched/bin/registry.terraform.io
│   └── getstackhead
│       └── acme
│           └── 1.5.0-patched
│               └── linux_amd64
│                   └── terraform-provider-acme_v1.5.0-patched
```

Reply:
This issue had nothing to do with the provider source address but with Terraform not following symlinks. According to the output  `only the base directory <plugin-dir> may be a symlink`, rules... are meant to be broken!

```
Initializing provider plugins...
- Finding hashicorp/cloudinit versions matching "1.0.0"...
2020/10/06 22:35:06 [WARN] Provider plugin search ignored symlink /nix/store/ws0i4lifcj0sk4x5zhxbqkyj29b5yyvp-tf-plugin-env/bin/terraform-provider-cloudinit_v1.0.0: only the base directory /nix/store/ws0i4lifcj0sk4x5zhxbqkyj29b5yyvp-tf-plugin-env/bin may be a symlink
2020/10/06 22:35:06 [WARN] Provider plugin search ignored symlink /nix/store/ws0i4lifcj0sk4x5zhxbqkyj29b5yyvp-tf-plugin-env/bin/terraform-provider-dns_v2.2.0: only the base directory /nix/store/ws0i4lifcj0sk4x5zhxbqkyj29b5yyvp-tf-plugin-env/bin may be a symlink
- Finding latest version of hashicorp/dns...

Error: Failed to query available provider packages

Could not retrieve the list of available versions for provider
hashicorp/cloudinit: provider registry.terraform.io/hashicorp/cloudinit was
not found in any of the search locations

- /nix/store/ws0i4lifcj0sk4x5zhxbqkyj29b5yyvp-tf-plugin-env/bin


Error: Failed to query available provider packages

Could not retrieve the list of available versions for provider hashicorp/dns:
provider registry.terraform.io/hashicorp/dns was not found in any of the
search locations

- /nix/store/ws0i4lifcj0sk4x5zhxbqkyj29b5yyvp-tf-plugin-env/bin
```

Reply:
Good investigation!

This seems to shift the implementation to the `withPlugins` function. Each plugin has to expose their attributes, and then `withPlugins` generates a trampoline script in the right location for each of the included providers.

Reply:
Fixed!
```tf
# main.tf
terraform {
  required_providers {
    acme = {
      source  = "getstackhead/acme"
      version = "1.5.0-patched"
    }
    auth0 = {
      source  = "nixpkgs/terraform-providers/auth0"
      version = "0.11.0"
    }
    cloudinit = {
      source  = "hashicorp/cloudinit"
      version = "1.0.0"
    }
    dns = {
      source  = "hashicorp/dns"
      version = "2.2.0"
    }
  }
}

provider "acme" {}
provider "auth0" {}
provider "cloudinit" {}
provider "dns" {}
```

```sh
nix-shell -I nixpkgs=$HOME/projects/nix/nixpkgs -p 'terraform_0_13.withPlugins (p : with p; [acme auth0 cloudinit dns])' --command "terraform init"

Initializing the backend...

Initializing provider plugins...
- Finding getstackhead/acme versions matching "1.5.0-patched"...
- Finding nixpkgs/terraform-providers/auth0 versions matching "0.11.0"...
- Finding hashicorp/cloudinit versions matching "1.0.0"...
- Finding hashicorp/dns versions matching "2.2.0"...
- Installing getstackhead/acme v1.5.0-patched...
- Installed getstackhead/acme v1.5.0-patched (unauthenticated)
- Installing nixpkgs/terraform-providers/auth0 v0.11.0...
- Installed nixpkgs/terraform-providers/auth0 v0.11.0 (unauthenticated)
- Installing hashicorp/cloudinit v1.0.0...
- Installed hashicorp/cloudinit v1.0.0 (unauthenticated)
- Installing hashicorp/dns v2.2.0...
- Installed hashicorp/dns v2.2.0 (unauthenticated)

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
```

![image](https://user-images.githubusercontent.com/1334474/95495381-40fd8400-0997-11eb-9b6c-9ad4f77a68d1.png)